chips/lowrisc/src/hmac.rs

```diff
@@ -0,0 +1,252 @@
+
+        regs.intr_enable
+            .modify(INTR_ENABLE::HMAC_DONE::CLEAR + INTR_ENABLE::HMAC_ERR::CLEAR);
+
+        if intrs.is_set(INTR_STATE::HMAC_DONE) {
+            self.client.map(|client| {
+                let digest = self.digest.take().unwrap();
+
+                for i in 0..8 {
+                    let d = regs.digest[i].get().to_ne_bytes();
+
+                    let idx = i * 4;
+
+                    digest[idx + 0] = d[3];
+                    digest[idx + 1] = d[2];
+                    digest[idx + 2] = d[1];
+                    digest[idx + 3] = d[0];
+                }
+
+                regs.intr_state.modify(INTR_STATE::HMAC_DONE::SET);
+
+                client.hash_done(Ok(()), digest);
+            });
+        } else if intrs.is_set(INTR_STATE::FIFO_FULL) {
+            // FIFO is full, we can't do anything
+        } else if intrs.is_set(INTR_STATE::HMAC_ERR) {
+            regs.intr_state.modify(INTR_STATE::HMAC_ERR::SET);
+
+            self.client.map(|client| {
+                client.hash_done(Err(ReturnCode::FAIL), self.digest.take().unwrap());
+            });
+        }
+    }
+}
+
+impl hil::digest::Digest<'a, [u8; 32]> for Hmac<'a> {
+    fn set_client(&'a self, client: &'a dyn digest::Client<'a, [u8; 32]>) {
+        self.client.set(client);
+    }
+
+    fn add_data(
+        &self,
+        data: LeasableBuffer<'static, u8>,
+    ) -> Result<usize, (ReturnCode, &'static mut [u8])> {
+        let regs = self.registers;
+
+        // Ensure the HMAC is setup
+        regs.cfg
+            .write(CFG::ENDIAN_SWAP::SET + CFG::SHA_EN::SET + CFG::DIGEST_SWAP::SET);
+
+        regs.cmd.modify(CMD::START::SET);
+
+        // Set the length and data index of the data to write
+        self.data_len.set(data.len());
+        self.data.set(Some(data));
+        self.data_index.set(0);
+
+        // Call the process function, this will start an async fill method
+        self.data_progress();
+
+        Ok(self.data_len.get())
+    }
+
+    fn run(
+        &'a self,
+        digest: &'static mut [u8; 32],
+    ) -> Result<(), (ReturnCode, &'static mut [u8; 32])> {
+        let regs = self.registers;
+
+        // Enable interrrupts
+        regs.intr_enable
+            .modify(INTR_ENABLE::HMAC_DONE::SET + INTR_ENABLE::HMAC_ERR::SET);
+
+        // Start the process
+        regs.cmd.modify(CMD::PROCESS::SET);
+
+        self.digest.set(Some(digest));
+
+        Ok(())
+    }
+
+    fn clear_data(&self) {
+        let regs = self.registers;
+
+        regs.cmd.modify(CMD::START::CLEAR);
+        regs.wipe_secret.set(1 as u32);
+    }
+}
+
+impl hil::digest::HMACSha256 for Hmac<'a> {
+    fn set_mode_hmacsha256(&self, key: &[u8; 32]) -> Result<(), ReturnCode> {
+        let regs = self.registers;
+
+        // Ensure the HMAC is setup
+        regs.cfg
+            .write(CFG::ENDIAN_SWAP::SET + CFG::SHA_EN::SET + CFG::DIGEST_SWAP::SET);
+
+        let key_len = core::cmp::min(key.len(), 8);
```

Comment:
Something I missed at first. `key` is now a fixed-size array instead of a slice. Comparing the length is useless as `key.len()` will always return 32.